mcorpc/authz_rego.go

```diff
@@ -0,0 +1,149 @@
+package mcorpc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/choria-io/go-choria/choria"
+	"github.com/choria-io/go-config"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/sirupsen/logrus"
+)
+
+type regoMatcher interface {
+	Set(caller string, actions string, facts string, classes string)
+}
+
+type regoPolicy struct {
+	cfg     *config.Config
+	req     *Request
+	agent   *Agent
+	log     *logrus.Entry
+	matcher regoMatcher
+}
+
+type regoPolicyPolicy struct {
+	caller  string
+	actions string
+	facts   string
+	classes string
+	log     *logrus.Entry
+}
+
+func (r *regoPolicyPolicy) Set(caller string, actions string, facts string, classes string) {
+	r.caller = caller
+	r.actions = actions
+	r.facts = facts
+	r.classes = classes
+}
+
+func regoPolicyAuthorize(req *Request, agent *Agent, log *logrus.Entry) (bool, error) {
+	logger := log.WithFields(logrus.Fields{
+		"authorizer": "regoPolicy",
+		"agent":      agent.Name(),
+		"request":    req.RequestID,
+	})
+
+	authz := &regoPolicy{
+		cfg:     agent.Config,
+		req:     req,
+		agent:   agent,
+		matcher: &regoPolicyPolicy{log: logger},
+		log:     logger,
+	}
+
+	log.Infof("Agent is %s", agent.Name())
+
+	return authz.authorize()
+}
+
+func (r *regoPolicy) authorize() (bool, error) {
+	policyFile, err := r.lookupPolicyFile()
+	if err != nil {
+		return false, err
+	}
+
+	pol, err := ioutil.ReadFile(policyFile)
+	if err != nil {
+		return false, err
+	}
+
+	module := string(pol)
+
+	query := rego.New(
+		rego.Query("data.choria.mcorpc.authpolicy.allow"),
+		rego.Module("//choria/mcorpc/policies/"+r.agent.meta.Name, module),
+		rego.Input(r.regoInputs()),
+	)
+
+	rs, err := query.Eval(context.Background())
+
+	switch err := err.(type) {
+	case nil:
+		break
+	case ast.Errors:
+		for _, e := range err {
+			r.log.Info("code: ", e.Code)
+			r.log.Info("row: ", e.Location.Row)
+			r.log.Info("filename: ", policyFile)
+		}
+	default:
+		return false, err
+	}
+
+	return rs[0].Expressions[0].Value.(bool), nil
+}
+
+func (r *regoPolicy) shouldUseDefault() bool {
+	enabled, err := choria.StrToBool(r.cfg.Option("plugin.regopolicy.enable_default", "n"))
+
+	if err != nil {
+		r.log.Debug("Parsing plugin.regopolicy.enable_default failed?:", err)
+		return false
+	}
+
+	return enabled
+}
+
+func (r *regoPolicy) lookupPolicyFile() (string, error) {
+	dir := filepath.Join(filepath.Dir(r.cfg.ConfigFile), "policies/rego")
+
+	regoPolicy := filepath.Join(dir, r.agent.Name()+".rego")
+
+	r.log.Debugf("Looking up rego policy in %s", regoPolicy)
+	if choria.FileExist((regoPolicy)) {
+		return regoPolicy, nil
+	}
+
+	if r.shouldUseDefault() {
```

Comment:
lets always support the default so drop this conditional and the shouldUseDefault function